### PR TITLE
Raise more detailed errors when calling IO inside defn

### LINKED ISSUE #731: Raise error message pointing to inspect_value and inspect_expr if someone tries to use IO inside defn

### PR DESCRIPTION
**Pull request by jonatanklosko**
Closes #731.

### CODE REVIEW

File: nx/lib/nx/defn/compiler.ex [hunk 1]
```diff
@@ -141,8 +141,16 @@ defmodule Nx.Defn.Compiler do
...
                 formatted = Exception.format_mfa(module, function, length(args))
 
-                reraise "cannot invoke #{formatted} inside defn because it was not defined with defn",
-                        stack
+                message =
+                  "cannot invoke #{formatted} inside defn because it was not defined with defn"
+
+                detail =
+                  case module do
+                    IO -> ". Use inspect_value/2 or inspect_expr/2 instead"
```
Comment (josevalim) [inline nx/lib/nx/defn/compiler.ex:149]: Let’s try to add a bit more information.  We can say,  if you want to print the runtime value of a tensor, use inspect_value. To print the tensor expression at compile time, use inspect_expr.

Comment (jonatanklosko) [reply nx/lib/nx/defn/compiler.ex]: Updated! Perhaps "compile time" is a bit confusing for someone knowing Elixir and new to Nx. Maybe we can word it differently, like "at compilation stage"?

File: nx/lib/nx/defn/compiler.ex [hunk 2]
```diff
@@ -141,8 +141,20 @@ defmodule Nx.Defn.Compiler do
...
-                reraise "cannot invoke #{formatted} inside defn because it was not defined with defn",
-                        stack
+                message =
+                  "cannot invoke #{formatted} inside defn because it was not defined with defn"
+
+                detail =
+                  case module do
+                    IO ->
+                      ". To print the runtime value of a tensor, use inspect_value/2. " <>
+                        "To print the tensor expression at compile time, use inspect_expr/2"
```
Comment (josevalim) [inline nx/lib/nx/defn/compiler.ex:151]:
```suggestion
                        "To print the tensor expression, use inspect_expr/2"
```

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**